src/data/real/ereal.lean

```diff
@@ -0,0 +1,159 @@
+negation is not, so we define negation. The main work is defining
+Sup : set ereal → ereal and proving that it is a Sup in all cases.
+Once we have, this, it is not hard to give a complete lattice
+structure on `ereal`.
+
+## Tags
+
+real, ereal, complete lattice
+-/
+
+/-- ereal : The type $$[-\infty,+\infty]$$ or `[-∞, ∞]` -/
+def ereal := with_bot (with_top ℝ)
+
+instance : linear_order ereal := by unfold ereal; apply_instance
+instance : lattice.order_bot ereal := by unfold ereal; apply_instance
+instance : lattice.order_top ereal := by unfold ereal; apply_instance
+
+/- neg -/
+
+/-- negation on ereal -/
+def ereal.neg : ereal → ereal
+| none := ⊤
+| (some none) := ⊥
+| (some (some x)) := ((↑-x : with_top ℝ) : with_bot (with_top ℝ))
+
+instance : has_neg ereal := ⟨ereal.neg⟩
+
+/-- if -a ≤ b then -b ≤ a on ereal -/
+theorem ereal.neg_le_of_neg_le : ∀ {a b : ereal} (h : -a ≤ b), -b ≤ a
+| none none h := by cases (lattice.le_bot_iff.1 h)
+| none (some b) h := by cases (lattice.top_le_iff.1 h); exact le_refl _
+| (some none) b h := lattice.le_top
+| (some (some a)) none h := by cases (lattice.le_bot_iff.1 h)
+| (some (some a)) (some none) h := lattice.bot_le
+| (some (some a)) (some (some b)) h := 
+begin
+  revert h,
+  change (((-a : ℝ) : with_top ℝ) : with_bot (with_top ℝ)) ≤ _ →
+    (((-b : ℝ) : with_top ℝ) : with_bot (with_top ℝ)) ≤ _,
+  unfold_coes, simpa using neg_le_of_neg_le,
+end
+
+/-- -a ≤ b ↔ -b ≤ a on ereal-/
+theorem ereal.neg_le {a b : ereal} : -a ≤ b ↔ -b ≤ a := ⟨ereal.neg_le_of_neg_le, ereal.neg_le_of_neg_le⟩
+
+/-- - -a = a on ereal -/
+theorem ereal.neg_neg : ∀ (a : ereal), - (- a) = a
+| none := rfl
+| (some none) := rfl
+| (some (some a)) := show (((- -a : ℝ) : with_top ℝ) : with_bot (with_top ℝ)) = (((a : ℝ) : with_top ℝ) : with_bot (with_top ℝ)),
+by simp [neg_neg a]
+
+/-- a ≤ -b → b ≤ -a on ereal -/
+theorem ereal.le_neg_of_le_neg {a b : ereal} (h : a ≤ -b) : b ≤ -a :=
+by rwa [←ereal.neg_neg b, ereal.neg_le, ereal.neg_neg]
+
+/-- The claim that a set of ereals has a supremum in ereal -/
+def has_Sup (X : set ereal) : Prop := ∃ l : ereal, is_lub X l
+
+local attribute [instance, priority 10] classical.prop_decidable
+
+/-- A set of ereals has a Sup in ereal -/
+theorem Sup_exists (X : set ereal) : has_Sup X :=
+  let Xoc : set (with_top ℝ) := λ x, X (↑x : with_bot _) in
+dite (Xoc = ∅) (λ h, ⟨⊥, ⟨
+    by
+    { rintro (⟨⟩|x) hx, exact le_refl none,
+      exfalso,
+      apply set.not_mem_empty x,
+      rw ←h,
+      exact hx,
+    },
+    λ u hu, lattice.bot_le⟩
+  ⟩) (λ h, dite (⊤ ∈ Xoc) (λ h2, ⟨⊤, ⟨λ _ _, lattice.le_top, λ x hx, hx _ h2⟩⟩) $
+    λ htop, 
+    let Xoo : set ℝ := λ (x : ℝ), Xoc (↑ x) in
+    begin
+    by_cases h2 : nonempty (upper_bounds Xoo),
+    { rcases h2 with ⟨b, hb⟩,
+      use (↑(↑(real.Sup Xoo : real) : with_top ℝ) : with_bot (with_top ℝ)),
+      split,
+      { rintros (⟨⟩|⟨⟩|x) hx,
+            exact lattice.bot_le,
```

Comment:
This use of indentation instead of focusing braces is not very mathlib.

Reply:
I've removed the longer ones.